aptos-move/aptos-workspace-server/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+    let shutdown = CancellationToken::new();
+    {
+        // TODO: Find a way to register the signal handler in a blocking manner without
+        //       waiting for it to trigger.
+        let shutdown = shutdown.clone();
+        tokio::spawn(async move {
+            tokio::signal::ctrl_c().await.unwrap();
+
+            println!("\nCtrl-C received. Shutting down services. This may take a while.\n");
+
+            shutdown.cancel();
+        });
+    }
+
+    // Phase 1: Start all services.
+    // Node
+    let (fut_node_api, fut_indexer_grpc, fut_node_finish) = services::node::start_node(test_dir)?;
+
+    let fut_node_api = make_shared(fut_node_api);
+    let fut_indexer_grpc = make_shared(fut_indexer_grpc);
+
+    // Faucet
+    let (fut_faucet, fut_faucet_finish) = services::faucet::start_faucet(
+        test_dir.to_owned(),
+        fut_node_api.clone(),
+        fut_indexer_grpc.clone(),
+    );
+
+    // Docker Network
+    let docker_network_name = format!("aptos-workspace-{}", instance_id);
+    let (fut_docker_network, fut_docker_network_clean_up) =
+        create_docker_network(shutdown.clone(), docker_network_name);
+
+    // Indexer part 1: postgres db
+    let (fut_postgres, fut_postgres_finish, fut_postgres_clean_up) =
+        services::postgres::start_postgres(
+            shutdown.clone(),
+            fut_docker_network.clone(),
+            instance_id,
+        );
+    let fut_postgres = make_shared(fut_postgres);
+
+    // Indexer part 2: processors
+    let (fut_all_processors_ready, fut_any_processor_finish) = start_all_processors(
+        fut_node_api.clone(),
+        fut_indexer_grpc.clone(),
+        fut_postgres.clone(),
+    );
+    let fut_all_processors_ready = make_shared(fut_all_processors_ready);
+
+    // Indexer part 3: indexer API
+    let (fut_indexer_api, fut_indexer_api_finish, fut_indexer_api_clean_up) = start_indexer_api(
+        instance_id,
+        shutdown.clone(),
+        fut_docker_network.clone(),
+        fut_postgres.clone(),
+        fut_all_processors_ready.clone(),
+    );
+
+    // Phase 2: Wait for all services to be up.
+    let all_services_up = async move {
+        tokio::try_join!(
+            fut_node_api.map_err(anyhow::Error::msg),
+            fut_indexer_grpc.map_err(anyhow::Error::msg),
+            fut_faucet,
+            fut_postgres.map_err(anyhow::Error::msg),
+            fut_all_processors_ready.map_err(anyhow::Error::msg),
+            fut_indexer_api,
+        )
+    };
+    let clean_up_all = async move {
+        eprintln!("Running shutdown steps");
+        fut_indexer_api_clean_up.await;
+        fut_postgres_clean_up.await;
+        fut_docker_network_clean_up.await;
+    };
+    tokio::select! {
+        _ = shutdown.cancelled() => {
+            clean_up_all.await;
+
+            return Ok(())
+        }
+        res = all_services_up => {
+            match res.context("one or more services failed to start") {
```

Comment:
can we know what services?

Reply:
We are supposed to, but there's some funky stuff happening with error propagation. 

My observation/theory is that when we have one service failing, other services that depend on it will likely short-cutting it thus preventing the original message from being shown. Let me double check the error propagation chain and make sure this gets fixed.

Reply:
Should be good enough for now, let's add a TODO to see if we can fix it later